File: src/cli/query/distinct.ts
import { Command, ExactArgs, ref } from "vipvot";
import { printJson, printError } from "../../lib/output.ts";
import { getMongoClient, closeAllClients } from "../../mongo/client.ts";
import { getDistinctValues } from "../../mongo/query.ts";
import { enhanceErrorMessage } from "../../lib/errors.ts";
import { parseJson } from "../../lib/parse-json.ts";
import { resolveConnectionAlias } from "../_globals.ts";

export function buildDistinctCommand(): Command {
  const filter = ref<string>("");

  const cmd = Command({
    use: "distinct <database> <collection> <field>",
    short: "Get distinct values for a field",
    args: ExactArgs(3),
    run: async (_cmd, args) => {
      const [database, collection, field] = args as [string, string, string];
      try {
        const { client } = await getMongoClient(resolveConnectionAlias());
        const filterDoc = filter.value ? parseJson(filter.value, "filter") : undefined;
        const values = await getDistinctValues(client, {
          dbName: database,
          collName: collection,
          field,
          filter: filterDoc,
        });
        printJson({ database, collection, field, values, count: values.length });
      } catch (err) {
        printError(
          err instanceof Error
            ? enhanceErrorMessage(err, { database, collection })
            : "Failed to get distinct values",
        );
      } finally {
        await closeAllClients();
      }
    },
  });

  cmd.flags().stringVarP(filter, "filter", "", "", "MongoDB query filter (JSON)");

  return cmd;
}

File: internal/cli/query/distinct.go
package query

import (
	"github.com/spf13/cobra"
	"go.mongodb.org/mongo-driver/v2/bson"

	"github.com/shhac/agent-mongo/internal/cli/shared"
	"github.com/shhac/agent-mongo/internal/mongo"
	"github.com/shhac/agent-mongo/internal/output"
)

func registerDistinct(parent *cobra.Command, globals func() *shared.GlobalFlags) {
	var filter string

	cmd := &cobra.Command{
		Use:   "distinct <database> <collection> <field>",
		Short: "Get distinct values for a field",
		Args:  cobra.ExactArgs(3),
		RunE: func(_ *cobra.Command, args []string) error {
			g := globals()
			ref := mongo.Ref{DB: args[0], Collection: args[1]}
			field := args[2]

			filterDoc, err := parseOptionalDoc(filter, "filter")
			if err != nil {
				return err
			}

			return shared.WithSessionRef(g, ref, func(ctx shared.SessionCtx) error {
				values, err := ctx.Session.DistinctValues(ctx.Ctx, ref, field, filterDoc)
				if err != nil {
					return err
				}
				return printDistinct(values, ref, field, filterDoc)
			})
		},
	}

	cmd.Flags().StringVar(&filter, "filter", "", "MongoDB query filter (JSON)")
	parent.AddCommand(cmd)
}

func printDistinct(values []any, ref mongo.Ref, field string, filter bson.D) error {
	var e echo
	e.str("field", field)
	e.doc("filter", filter)
	return output.PrintResultWithMeta(map[string]any{
		"database":   ref.DB,
		"collection": ref.Collection,
		"field":      field,
		"values":     values,
		"count":      len(values),
	}, e.meta())
}
